File: src/main/java/com/example/finsentinel/service/trading/uta/OkxBroker.java
package com.example.finsentinel.service.trading.uta;

import com.example.finsentinel.service.okx.OkxTradingEngine;
import com.example.finsentinel.service.trading.engine.AccountInfo;
import com.example.finsentinel.service.trading.engine.OrderRequest;
import com.example.finsentinel.service.trading.engine.OrderResult;
import com.example.finsentinel.service.trading.engine.PositionInfo;

import java.util.EnumSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;

/**
 * IBroker adapter for OKX crypto derivatives trading.
 * Supports CRYPTO and PERP security types.
 * Not a Spring bean — created by BrokerRegistry.
 */
public class OkxBroker implements IBroker {

    private final OkxTradingEngine engine;

    public OkxBroker(OkxTradingEngine engine) {
        this.engine = Objects.requireNonNull(engine, "OkxTradingEngine must not be null");
    }

    @Override
    public String brokerId() {
        return "okx";
    }

    @Override
    public String displayName() {
        return "OKX (Crypto Derivatives)";
    }

    @Override
    public Set<SecurityType> supportedSecurityTypes() {
        // OKX only handles perpetual swaps via UTA routing.
        // Crypto spot uses CCXT slash format (BTC/USDT) which OKX rejects —
        // spot orders should route to CcxtBroker instead.
        return EnumSet.of(SecurityType.PERP);
    }

    @Override
    public Set<BrokerCapability> capabilities() {
        return EnumSet.of(
                BrokerCapability.SPOT_TRADING,
                BrokerCapability.PERPETUAL_SWAP,
                BrokerCapability.MARGIN_TRADING,
                BrokerCapability.MARKET_DATA,
                BrokerCapability.ORDER_MANAGEMENT);
    }

    @Override
    public OrderResult placeOrder(Contract contract, OrderRequest request) {
        var nativeRequest = new OrderRequest(
                contract.toEngineSymbol(), request.side(), request.type(),
                request.qty(), request.notional(), request.price(),
                request.stopPrice(), request.timeInForce(), request.reduceOnly());
        return engine.placeOrder(nativeRequest);
    }

    @Override
    public List<PositionInfo> getPositions() {
        return engine.getPositions();
    }

    @Override
    public List<OrderResult> getOrders() {
        return engine.getOrders();
    }

    @Override
    public AccountInfo getAccount() {
        return engine.getAccount();
    }

    @Override
    public boolean cancelOrder(String orderId) {
        return engine.cancelOrder(orderId);
    }

    /** Exposes the underlying engine for OKX-specific operations (funding rate, leverage). */
    public OkxTradingEngine engine() {
        return engine;
    }
}

File: apps/api/src/trading/brokers/okx.broker.ts
import {
  SecurityType,
  BrokerCapability,
  Contract,
} from '@finsentinel/shared';
import type { IBroker } from '../interfaces/broker';
import type { OkxTradingEngine } from '../../okx/okx-trading.engine';
import type {
  OrderRequest,
  OrderResult,
  PositionInfo,
  AccountInfo,
  MarketClock,
} from '../interfaces/types';

/**
 * OKX broker capabilities -- perpetual swaps and futures.
 */
const OKX_CAPABILITIES: Set<BrokerCapability> = new Set([
  BrokerCapability.PERPETUAL_SWAP,
  BrokerCapability.MARGIN_TRADING,
]);

/**
 * OkxBroker -- wraps OkxTradingEngine with Contract-aware symbol conversion.
 *
 * NOT an Injectable. Created by BrokerRegistry at runtime.
 *
 * OKX broker adapter:
 * - Supports PERP and FUTURE security types
 * - Converts contract to engine-native symbol via Contract.toEngineSymbol()
 * - Delegates all operations to the underlying OkxTradingEngine
 */
export class OkxBroker implements IBroker {
  constructor(private readonly _engine: OkxTradingEngine) {}

  // -- Identity ---------------------------------------------------------------

  brokerId(): string {
    return 'okx';
  }

  displayName(): string {
    return 'OKX (Crypto Derivatives)';
  }

  // -- Capability checks ------------------------------------------------------

  supportedSecurityTypes(): Set<SecurityType> {
    return new Set([SecurityType.PERP, SecurityType.FUTURE]);
  }

  capabilities(): Set<BrokerCapability> {
    return OKX_CAPABILITIES;
  }

  canHandle(contract: Contract): boolean {
    return (
      contract.secType === SecurityType.PERP ||
      contract.secType === SecurityType.FUTURE
    );
  }

  // -- Order execution --------------------------------------------------------

  async placeOrder(
    contract: Contract,
    request: OrderRequest,
  ): Promise<OrderResult> {
    const adapted: OrderRequest = {
      ...request,
      symbol: contract.toEngineSymbol(),
    };

    return this._engine.placeOrder(adapted);
  }

  // -- Delegated operations ---------------------------------------------------

  async getPositions(): Promise<PositionInfo[]> {
    return this._engine.getPositions();
  }

  async getOrders(): Promise<OrderResult[]> {
    return this._engine.getOrders();
  }

  async getAccount(): Promise<AccountInfo> {
    return this._engine.getAccount();
  }

  async cancelOrder(orderId: string): Promise<boolean> {
    return this._engine.cancelOrder(orderId);
  }

  async syncOrders(): Promise<OrderResult[]> {
    return this._engine.syncOrders!();
  }

  async getMarketClock(): Promise<MarketClock> {
    return this._engine.getMarketClock!();
  }

  async searchContracts(_query: string): Promise<Contract[]> {
    return [];
  }

  // -- Engine access ----------------------------------------------------------

  /** Expose the underlying engine for direct access when needed. */
  engine(): OkxTradingEngine {
    return this._engine;
  }
}
